src/payload.rs

```diff
@@ -0,0 +1,516 @@
+                return (num_read, num_written, clocks_overflowed);
+            }
+            // Push will not panic, as free capacity is known to be at least 1
+            output.push(cur_entry);
+            num_written += 1;
+            num_read += 1;
+        }
+    }
+    (num_written, num_read, clocks_overflowed)
+}
+
+/// Push frontier clocks to output
+/// Note: will panic if not enough space to fit all clocks
+fn write_clocks<C, O>(frontier_clocks: &mut C, output: &mut O)
+where
+    C: FrontierClocks,
+    O: PayloadOutput,
+{
+    for c in frontier_clocks.as_slice() {
+        let (id_entry, clock_entry) = LogEntry::clock(*c);
+        output.push(id_entry);
+        output.push(clock_entry);
+    }
+}
+
+/// Append a "logs missed" event to the output
+///
+/// Note: assumes free capacity of output is at least 2
+fn push_log_items_missed_event<O>(output: &mut O, n: u32)
+where
+    O: PayloadOutput,
+{
+    let (event_id_entry, payload_entry) =
+        LogEntry::event_with_payload(EventId::EVENT_LOG_ITEMS_MISSED, n);
+    output.push(event_id_entry);
+    output.push(payload_entry);
+}
+
+#[cfg(test)]
+mod tests {
+    use super::*;
+    use crate::{ProbeEpoch, ProbeTicks};
+    use core::mem::MaybeUninit;
+
+    fn log_to_bytes<'a, 'b>(log: &'b [LogEntry], out_slice: &'a mut [u8]) -> usize {
+        let mut out = PayloadOutputByteSlice::new(out_slice);
+        let mut n = 0;
+        for item in log {
+            out.push(*item);
+            n += 1;
+        }
+        drop(out);
+        n * size_of::<LogEntry>()
+    }
+
+    /// Write a simple payload and merge the clock getting reported
+    #[test]
+    fn simple_payload() {
+        let probe_id = ProbeId::new(1).unwrap();
+        let mut frontier_clocks_arr = [MaybeUninit::<LogicalClock>::uninit(); 2];
+        let mut frontier_clocks = FixedSliceVec::new(&mut frontier_clocks_arr);
+        FrontierClocksFSV(&mut frontier_clocks)
+            .merge_clock(probe_id, 0)
+            .unwrap();
+
+        let log = [
+            PossiblyMissed::Entry(LogEntry::event(EventId::new(1).unwrap())),
+            PossiblyMissed::Entry(LogEntry::event(EventId::new(2).unwrap())),
+            PossiblyMissed::Entry(
+                LogEntry::clock(LogicalClock {
+                    id: probe_id,
+                    ticks: ProbeTicks(1),
+                    epoch: ProbeEpoch(0),
+                })
+                .0,
+            ),
+            PossiblyMissed::Entry(
+                LogEntry::clock(LogicalClock {
+                    id: probe_id,
+                    ticks: ProbeTicks(1),
+                    epoch: ProbeEpoch(0),
+                })
+                .1,
+            ),
+        ];
+
+        let mut payload = [0u8; 128];
+        let (num_written, num_read, did_clocks_overflow) = write_report_payload(
+            log.iter().copied(),
+            &mut FrontierClocksFSV::new(&mut frontier_clocks),
```

Comment:
Why are we making a new instance here rather than holding on to the instance created earlier in the test?